test/dsa_no_digest_size_test.c

```diff
@@ -0,0 +1,240 @@
+	0xA7, 0x0B, 0x6E, 0x96, 0xB8, 0xC8, 0xCB, 0x38, 0x35, 0x12, 
+	0x20, 0x79, 0xA5, 0x08, 0x28, 0x35, 0x5C, 0xBC, 0x52, 0x16, 
+	0xAF, 0x52, 0xBA, 0x0F, 0xC3, 0xB1, 0x63, 0x12, 0x27, 0x0B, 
+	0x74, 0xA4, 0x47, 0x43, 0xD6, 0x30, 0xB8, 0x9C, 0x2E, 0x40, 
+	0x14, 0xCD, 0x99, 0x7F, 0xE8, 0x8E, 0x37, 0xB0, 0xA9, 0x3F, 
+	0x54, 0xE9, 0x66, 0x22, 0x61, 0x4C, 0xF8, 0x49, 0x03, 0x57, 
+	0x14, 0x32, 0x1D, 0x37, 0x3D, 0xE2, 0x92, 0xF8, 0x8E, 0xA0, 
+	0x6A, 0x66, 0x63, 0xF0, 0xB0, 0x6E, 0x07, 0x2B, 0x3D, 0xBF, 
+	0xD0, 0x84, 0x6A, 0xAA, 0x1F, 0x30, 0x77, 0x65, 0xE5, 0xFC, 
+	0xF5, 0xEC, 0x55, 0xCE, 0x73, 0xDB, 0xBE, 0xA7, 0x8D, 0x3A, 
+	0x9F, 0x7A, 0xED, 0x4F, 0xAF, 0xA2, 0x80, 0x4C, 0x30, 0x9E, 
+	0x28, 0x49, 0x65, 0x40, 0xF0, 0x03, 0x45, 0x56, 0x99, 0xA2, 
+	0x93, 0x1B, 0x9C, 0x46, 0xDE, 0xBD, 0xA8, 0xAB, 0x5F, 0x90, 
+	0x3F, 0xB7, 0x3F, 0xD4, 0x6F, 0x8D, 0x5A, 0x30, 0xE1, 0xD4, 
+	0x63, 0x3A, 0x6A, 0x7C, 0x8F, 0x24, 0xFC, 0xD9, 0x14, 0x28, 
+	0x09, 0xE4, 0x84, 0x4E, 0x17, 0x43, 0x56, 0xB8, 0xD4, 0x4B, 
+	0xA2, 0x29, 0x45, 0xD3, 0x13, 0xF0, 0xC2, 0x76, 0x9B, 0x01, 
+	0xA0, 0x80, 0x6E, 0x93, 0x63, 0x5E, 0x87, 0x24, 0x20, 0x2A, 
+	0xFF, 0xBB, 0x9F, 0xA8, 0x99, 0x6C, 0xA7, 0x9A, 0x00, 0xB9, 
+	0x7D, 0xDA, 0x66, 0xC9, 0xC0, 0x72, 0x72, 0x22, 0x0F, 0x1A, 
+	0xCC, 0x23, 0xD9, 0xB7, 0x5F, 0x1B
+    };
+    DSA *dsa = DSA_new();
+    BIGNUM *p, *q, *g;
+
+    if (dsa == NULL)
+        return NULL;
+    if (!DSA_set0_pqg(dsa, p = BN_bin2bn(dsap_2048, sizeof(dsap_2048), NULL),
+                           q = BN_bin2bn(dsaq_2048, sizeof(dsaq_2048), NULL),
+                           g = BN_bin2bn(dsag_2048, sizeof(dsag_2048), NULL))) {
+        DSA_free(dsa);
+        BN_free(p);
+        BN_free(q);
+        BN_free(g);
+        return NULL;
+    }
+    return dsa;
+}
+
+static int genkeys(void)
+{
+    if (!TEST_ptr(dsakey = load_dsa_params()))
+        return 0;
+
+    if (!TEST_int_eq(DSA_generate_key(dsakey), 1))
+        return 0;
+
+    return 1;
+}
+
+static int sign_and_verify(int len)
+{
+    /*
+     * Per FIPS 186-4, the hash is recommended to be the same length as q.
+     * If the hash is longer than q, the leftmost N bits are used; if the hash
+     * is shorter, then we left-pad (see appendix C.2.1).
+     */
+    size_t sigLength;
+    int digestlen = BN_num_bytes(DSA_get0_q(dsakey));
+    int ok = 0;
+
+    unsigned char *dataToSign = OPENSSL_malloc(len);
+    unsigned char *paddedData = OPENSSL_malloc(digestlen);
+    unsigned char *signature = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    if (!TEST_ptr(dataToSign) ||
+        !TEST_ptr(paddedData) ||
+        !TEST_int_eq(RAND_bytes(dataToSign, len), 1))
+        goto end;
+
+    memset(paddedData, 0, digestlen);
+    if (len > digestlen)
+        memcpy(paddedData, dataToSign, digestlen);
+    else
+        memcpy(paddedData + digestlen - len, dataToSign, len);
+    
+    if (!TEST_ptr(pkey = EVP_PKEY_new())) goto end;
```

Comment:
sorry to do this, but the "goto" part should be on a separate line.  for all of them.

Reply:
Oops. For future reference, is there an openssl style check tool I can use to avoid these back-and-forths? :)

Reply:
There is a script in the util directory, but it's not perfect.  The style guide is online, at https://www.openssl.org/policies/codingstyle.html